### PR TITLE
Convert `uniXXXX` glyph names to proper ones when building the `charCodeToGlyphId` map for TrueType fonts (bug 1132849, issue 6893, issue 6894)

### LINKED ISSUE #6893: Characters displayed as squares
*I have tested this with the preview from PR #6891 as that makes the PDF render properly.*

I already noticed this before with Scribus-generated PDF files, but now I have been able to make a reduced test case. Opening the file below with the preview at http://107.21.233.14:8877/20a6d5dd16fdaa6/web/viewer.html makes most characters display properly, but some display as a square, in particular `!`, `"` and `-`. Adobe Acrobat Reader DC has no problem with these glyphs.

Below is a reduced test case for inclusion in a PR that fixes this issue. It might be possible to reduce it even further.
[reduced-characters.pdf](https://github.com/mozilla/pdf.js/files/94478/reduced-characters.pdf)

### LINKED ISSUE #6894: Font styles are not applied
*I have tested this with the preview from PR #6891 as that makes the PDF render properly.*

Opening the file below with the preview at http://107.21.233.14:8877/20a6d5dd16fdaa6/web/viewer.html makes all characters display properly, but the font styles (bold and italic) are not applied. Adobe Acrobat Reader DC has no problem with this. The PDF is generated by Scribus 1.5.0svn.

Below is a reduced test case for inclusion in a PR that fixes this issue. It might be possible to reduce it even further.
[reduced-font-styles.pdf](https://github.com/mozilla/pdf.js/files/94490/reduced-font-styles.pdf)

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Some PDF generators, e.g. Scribus PDF, use improper `uniXXXX` glyph names which breaks the glyph mapping. We can avoid this by converting them to "standard" glyph names instead.

Fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1132849.
Fixes #6893.
Fixes #6894.

/cc @timvandermeij

### CODE REVIEW

**Comment by timvandermeij** [comment]
This is awesome! I can confirm that this fixes both issues I filed. The original and the reduced test case from the Bugzilla issue already seem to be fixed for me with the current master (on Arch Linux when comparing PDF.js with Okular), but nevertheless it's good to include it as a reduced test case.

@brendandahl Could you please review this?

**Comment by Snuffleupagus** [comment]
> The original and the reduced test case from the Bugzilla issue already seem to be fixed for me with the current master (on Arch Linux when comparing PDF.js with Okular), but nevertheless it's good to include it as a reduced test case.

For me there's actually a difference on Windows (with HWA on), which is why I included it.

render_diff

---

Although there's now a couple of somewhat similar test-cases, which might be unnecessary. The already existing `issue6889.pdf` could probably be replaced with `issue6894.pdf` in this patch, since those two appear to be using the exact same font as (but the new one covers more issues).

However, I didn't want to remove any existing tests in this patch.

**Comment by Snuffleupagus** [comment]
I've update the patch slightly, to only attempt this conversion for the `this.differences` case. Since otherwise we just read from one of the standard encodings, which we know don't suffer from this issue.

/botio test

**Comment by brendandahl** [comment]
This is looking good. I was curious how other PDF viewers handle this and found https://pdfium.googlesource.com/pdfium/+/master/core/src/fxge/freetype/fx_freetype.cpp#56 . Seems like it would be a good idea to make a general function for getting the unicode from glyph name. There function handles names like 'u123456' and a few other cases. Now I wonder if we have any broken pdfs with names like the above.

**Comment by Snuffleupagus** [comment]
> Seems like it would be a good idea to make a general function for getting the unicode from glyph name.

Sure, that sounds fine! The only one I can remember seeing in practice is `uniXXXX`, but I suppose it can't hurt to support the `uXXXX/uXXXXXX` case as well.

@brendandahl How closely do you want that function to mimic what the FreeType code does?
Should we enforce only upper-case hexadecimal chars, to prevent accidentally matching things that are not actually hex numbers?

**Comment by Snuffleupagus** [comment]
I've added a `getUnicodeForGlyph` helper function, similarly to the FreeType one, which is used in various places to do the `glyphName -> Unicode` conversion.
I kept the `recoverGlyphName`, which is using `getUnicodeForGlyph` internally, since we may need both the standard/non-standard glyph names when creating the `charCodeToGlyphId` map for TrueType fonts.

@brendandahl Is this what you had in mind?

/botio-linux preview

File: src/core/unicode.js [hunk 1]
```diff
@@ -65,6 +65,34 @@
...
+    }
+    // Try to recover valid Unicode values from 'uniXXXX'/'uXXXX{XX}' glyphs.
+    if (name[0] === 'u') {
+      var nameLen = name.length, hexStr;
+
+      if (name[1] === 'n' && name[2] === 'i' && nameLen === 7) { // 'uniXXXX'
+        hexStr = name.substr(3);
+      } else if (nameLen >= 5 && nameLen <= 7) { // 'uXXXX{XX}'
+        hexStr = name.substr(1);
+      }
```
Comment (brendandahl) [inline src/core/unicode.js]: It's a bit easier to follow if we add an `else` and return -1 early here.

File: src/core/unicode.js [hunk 2]
```diff
@@ -65,6 +65,34 @@
...
+      return unicode;
+    }
+    if (!name) {
+      return -1;
+    }
+    // Try to recover valid Unicode values from 'uniXXXX'/'uXXXX{XX}' glyphs.
+    if (name[0] === 'u') {
+      var nameLen = name.length, hexStr;
+
+      if (name[1] === 'n' && name[2] === 'i' && nameLen === 7) { // 'uniXXXX'
```
Comment (brendandahl) [inline src/core/unicode.js]: This isn't in a tight loop or anything, but it's best to do the cheapest check first (nameLen === 7) so the rest are short circuited.

File: src/core/unicode.js [hunk 3]
```diff
@@ -65,6 +65,34 @@
...
+        hexStr = name.substr(3);
+      } else if (nameLen >= 5 && nameLen <= 7) { // 'uXXXX{XX}'
+        hexStr = name.substr(1);
+      }
+      // Check for upper-case hexadecimal characters, to avoid false positives.
+      if (hexStr && hexStr === hexStr.toUpperCase()) {
```
Comment (brendandahl) [inline src/core/unicode.js]: With the above you can remove the first hexStr condition.

**Comment by brendandahl** [comment]
> @brendandahl Is this what you had in mind?

Yes, looks very good. Just the minor nits above.

**Comment by Snuffleupagus** [comment]
> Yes, looks very good. Just the minor nits above.

Thanks for the review, I've addressed the comments.

/botio test

**Comment by brendandahl** [comment]
r+
/botio makeref

**Pull request merged**